src/compiler/scanner.ts

```diff
@@ -1105,14 +1110,69 @@ export function createScanner(languageVersion: ScriptTarget,
             break;
         }
         if (text.charCodeAt(pos - 1) === CharacterCodes._) {
+            tokenFlags |= TokenFlags.ContainsInvalidSeparator;
             error(Diagnostics.Numeric_separators_are_not_allowed_here, pos - 1, 1);
         }
         return result + text.substring(start, pos);
     }
 
+    // Extract from Section 12.9.3
+    // NumericLiteral ::=
+    //     | DecimalLiteral
+    //     | DecimalBigIntegerLiteral
+    //     | NonDecimalIntegerLiteral 'n'?
+    //     | LegacyOctalIntegerLiteral
+    // DecimalBigIntegerLiteral ::=
+    //     | '0n'
+    //     | [1-9] DecimalDigits? 'n'
+    //     | [1-9] '_' DecimalDigits 'n'
+    // DecimalLiteral ::=
+    //     | DecimalIntegerLiteral? '.' DecimalDigits? ExponentPart?
+    //     | '.' DecimalDigits ExponentPart?
+    //     | DecimalIntegerLiteral ExponentPart?
+    // DecimalIntegerLiteral ::=
+    //     | '0'
+    //     | [1-9] '_'? DecimalDigits
+    //     | NonOctalDecimalIntegerLiteral
+    // LegacyOctalIntegerLiteral ::= '0' [0-7]+
+    // NonOctalDecimalIntegerLiteral ::= '0' [0-7]* [89] [0-9]*
     function scanNumber(): { type: SyntaxKind, value: string } {
-        const start = pos;
-        const mainFragment = scanNumberFragment();
+        let start = pos;
+        let mainFragment: string;
+        if (text.charCodeAt(pos) === CharacterCodes._0) {
+            pos++;
+            if (text.charCodeAt(pos) === CharacterCodes._) {
+                tokenFlags |= TokenFlags.ContainsSeparator | TokenFlags.ContainsInvalidSeparator;
+                error(Diagnostics.Numeric_separators_are_not_allowed_here, pos, 1);
+                // treat it as a normal number literal
+                pos--;
+                mainFragment = scanNumberFragment();
+            }
+            // Separators are not allowed in the below cases
+            else if (!scanDigits()) {
+                // NonOctalDecimalIntegerLiteral, emit error later
+                // Separators in decimal and exponent parts are still allowed according to the spec
+                tokenFlags |= TokenFlags.ContainsLeadingZero;
+                mainFragment = "" + +tokenValue;
+            }
+            else if (!tokenValue) {
+                // a single zero
+                mainFragment = "0";
+            }
+            else {
+                // LegacyOctalIntegerLiteral
+                tokenValue = "" + parseInt(tokenValue, 8);
+                tokenFlags |= TokenFlags.Octal;
+                const withMinus = token === SyntaxKind.MinusToken;
+                const literal = (withMinus ? "-" : "") + "0o" + (+tokenValue).toString(8);
+                start -= +withMinus;

```

Comment:
minor: `if (withMinus) start--` avoids a maybe (?) slow boolean->numeric conversion and is also way less surprising (I typically expect only strings containing numbers to be converted to numbers, and my brain is broken by learning that True coerces to -1 in Visual Basic [1] many many years ago, so I always have trouble reading boolean->numeric math.)

[1] -1 is all 1s in twos-complement binary representation, you see. It almost makes sense!

Reply:
Well, at least `Ternary.True` is still -1 😂